### PR TITLE
Connector migration to new semantic impl

### PR DESCRIPTION
**Pull request by lankavitharana**

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java [hunk 1]
```diff
@@ -2562,52 +2561,52 @@ private void retryTransaction(int transactionId, int startOfAbortIP) {
...
-        // TODO Improve following two lines
-        this.constPool = calleeSF.packageInfo.getConstPoolEntries();
-        this.code = calleeSF.packageInfo.getInstructions();
-        ip = defaultWorkerInfo.getCodeAttributeInfo().getCodeAddrs();
-
-        // Invoke other workers
-        BLangVMWorkers.invoke(programFile, callableUnitInfo, callerSF, argRegs);
-
-    }
+//    public void invokeActionCallableUnit(ActionInfo callableUnitInfo, FunctionCallCPEntry funcCallCPEntry) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java]: Shall we delete this code? No need to keep them

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/codegen/ProgramFileReader.java [hunk 2]
```diff
@@ -430,19 +430,19 @@ private void readConnectorInfoEntries(DataInputStream dataInStream,
...
-                int value = dataInStream.readInt();
-                methodTableInteger.put(new Integer(key), new Integer(value));
-            }
-
-            connectorInfo.setMethodTableIndex(methodTableInteger);
-
-            boolean isFilterConnector = dataInStream.readBoolean();
-            connectorInfo.setFilterConnector(isFilterConnector);
+//            Map<Integer, Integer> methodTableInteger = new HashMap<>();
+//            int count = dataInStream.readInt();
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/codegen/ProgramFileReader.java]: Let's remove all these commented lines

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java [hunk 3]
```diff
@@ -775,7 +787,59 @@ public void visit(BLangInvocation iExpr) {
             } else {
                 emit(InstructionCodes.CALL, funcRefCPIndex, funcCallCPIndex);
             }
+        } else if (iExpr.expr != null) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: You can visit BLangActionInvocation directly in the code generator and generate code.

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -471,11 +497,20 @@ public void visit(BLangInvocation iExpr) {
...
-                // TODO 
+                List<BLangExpression> actionArgExprs = new ArrayList<>(iExpr.argExprs);
+                actionArgExprs.add(0, iExpr.expr);
+                result = new BLangActionInvocation(iExpr.pos, actionArgExprs, iExpr.expr, iExpr.symbol, iExpr.types);
                 break;
         }
     }
 
+    public void visit(BLangConnectorInit connectorInitExpr) {
+        BLangConnectorInit genCIExpr = connectorInitExpr;
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Why do we need this genCIExpr here?

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -471,11 +497,20 @@ public void visit(BLangInvocation iExpr) {
                 result = new BLangFunctionInvocation(iExpr.pos, argExprs, iExpr.symbol, iExpr.types);
                 break;
             case TypeTags.CONNECTOR:
-                // TODO 
+                List<BLangExpression> actionArgExprs = new ArrayList<>(iExpr.argExprs);
+                actionArgExprs.add(0, iExpr.expr);
+                result = new BLangActionInvocation(iExpr.pos, actionArgExprs, iExpr.expr, iExpr.symbol, iExpr.types);
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Use this new expression in the code generator

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 6]
```diff
@@ -268,8 +279,23 @@ public void visit(BLangAction actionNode) {
         BInvokableSymbol actionSymbol = Symbols
                 .createActionSymbol(Flags.asMask(actionNode.flagSet), names.fromIdNode(actionNode.name),
                         env.enclPkg.symbol.pkgID, null, env.scope.owner);
+        // Create and keep parameter node to inject as action first parameter later
+        BLangVariable connectorParam = (BLangVariable) TreeBuilder.createVariableNode();
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: actionNode.connector  rename this to actionNode.receiver :)

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 7]
```diff
@@ -268,8 +279,23 @@ public void visit(BLangAction actionNode) {
...
+        connectorParam.pos = env.node.pos;
+        connectorParam.setName(this.createIdentifier(Names.CONNECTOR.getValue()));
+        BLangUserDefinedType connectorType = (BLangUserDefinedType) TreeBuilder.createUserDefinedTypeNode();
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Do we need to keep the connector variable node here? We just need the connector symbol right?

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 8]
```diff
@@ -493,6 +526,32 @@ private void defineInvokableSymbolParams(BLangInvokableNode invokableNode, BInvo
         symbol.type = new BInvokableType(paramTypes, retTypes, null);
     }
 
+    private void defineConnectorSymbol(BLangConnector connectorNode, BTypeSymbol connectorSymbol,
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Do we need this method at all? Can we just put this in the Connector node visit method?

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 9]
```diff
@@ -493,6 +526,32 @@ private void defineInvokableSymbolParams(BLangInvokableNode invokableNode, BInvo
...
+                                       SymbolEnv connectorEnv) {
+        connectorNode.symbol = connectorSymbol;
+        defineSymbol(connectorNode.pos, connectorSymbol);
+        connectorEnv.scope = connectorSymbol.scope;
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: This line is redundant 

connectorEnv.scope = connectorSymbol.scope;

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 10]
```diff
@@ -531,26 +590,38 @@ private void defineConnectorInitFunction(BLangConnector connector) {
         BLangUserDefinedType connectorType = (BLangUserDefinedType) TreeBuilder.createUserDefinedTypeNode();
         connectorType.pos = connector.pos;
         connectorType.typeName = connector.name;
+        connectorType.pkgAlias = (BLangIdentifier) TreeBuilder.createIdentifierNode();
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: This should happen at the BLangPackageBuilder level

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java [hunk 11]
```diff
@@ -318,8 +330,8 @@ public void visit(BLangUserDefinedType userDefinedTypeNode) {
         // 2) lookup the typename in the package scope returned from step 1.
         // 3) If the symbol is not found, then lookup in the root scope. e.g. for types such as 'error'
 
-        BSymbol pkgSymbol = resolvePkgSymbol(userDefinedTypeNode.pos, this.env,
-                names.fromIdNode(userDefinedTypeNode.pkgAlias));
+        BSymbol pkgSymbol = resolvePkgSymbol(userDefinedTypeNode.pos, this.env, names
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java]: This formatting is not correct IMO

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangNodeVisitor.java [hunk 12]
```diff
@@ -277,6 +278,10 @@ public void visit(BLangInvocation invocationExpr) {
         throw new AssertionError();
     }
 
+    public void visit(BLangConnectorInit connectorInitExpr) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangNodeVisitor.java]: Put BLangActionInvocation here

Comment (lankavitharana) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/BLangNodeVisitor.java]: done

File: modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 10 above)
Comment (lankavitharana) [reply modules/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: I'm not clear on this, how do we do that in package builder level?

**Pull request merged**